src/profiling.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License.  You may obtain a copy
+// of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+#[cfg(feature = "profiling")] use cpuprofiler::PROFILER;
+use failure::Fallible;
+use std::path::Path;
+
+/// Facade for `cpuprofiler::PROFILER` to cope with the optional gperftools dependency and to
+/// ensure profiling stops on `drop`.
+pub struct ScopedProfiler {}
+
+impl ScopedProfiler {
+    #[cfg(not(feature = "profiling"))]
+    fn real_start<P: AsRef<Path>>(_path: P) -> Fallible<ScopedProfiler> {
+        Err(format_err!("Compile-time \"profiling\" feature not enabled"))
+    }
+
+    #[cfg(feature = "profiling")]
+    fn real_start<P: AsRef<Path>>(path: P) -> Fallible<ScopedProfiler> {
+        let path_str = match path.as_ref().to_str() {
+            Some(path_str) => path_str,
+            None => return Err(format_err!("Invalid path {}", path.as_ref().display())),
```

Comment:
Don't need as_ref here.

Reply:
Looks like I do. If I remove it, the compiler complains that `P` doesn't have a `display()` method. And if I remove the `display()` call, then it complains about this not being printable with the default formatter.

Reply:
Oh, right, P is AsRef. I feel like you want `let path = path.as_ref()`, but it's not critical.

Reply:
Ah, that's a good idea. Done.